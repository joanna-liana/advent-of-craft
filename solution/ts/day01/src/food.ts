type Maybe<T> = NonNullable<T> | null;
type DateProvider = () => Date;

export class Food {
    constructor(
        public readonly expirationDate: Date,
        public readonly approvedForConsumption: boolean,
        public readonly inspectorId: Maybe<string>
    ) {
    }

    isEdible(now: DateProvider): boolean {
        return this.isFresh(now) &&
            this.canBeConsumed() &&
            this.hasBeenInspected();
    }

    private isFresh = (now: () => Date) => this.expirationDate > now();
    private canBeConsumed = () => this.approvedForConsumption;
    private hasBeenInspected = () => this.inspectorId !== null;
}