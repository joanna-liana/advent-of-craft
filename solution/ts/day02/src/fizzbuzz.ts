const min = 0;
const max = 100;
const fizz = 3;
const buzz = 5;
const fizzBuzz = 15;

export function fizzbuzz(input: number): string | number {
    if (isOutOfRange(input)) {
        throw new Error("Input is out of range")
    }
    return convertSafely(input);
}

function convertSafely(input: number): string | number {
    if (is(fizzBuzz, input)) {
        return "FizzBuzz";
    }
    if (is(fizz, input)) {
        return "Fizz";
    }
    if (is(buzz, input)) {
        return "Buzz";
    }
    return input;
}

function is(divisor: number, input: number): boolean {
    return input % divisor === 0;
}

function isOutOfRange(input: number): boolean {
    return input <= min || input > max;
}