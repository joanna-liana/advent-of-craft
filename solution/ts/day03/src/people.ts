export enum PetType {
    Cat,
    Dog,
    Hamster,
    Turtle,
    Bird,
    Snake
}

export class Pet {
    constructor(
        public readonly type: PetType,
        public readonly name: string,
        public readonly age: number
    ) {
    }
}

export class Person {
    constructor(
        public readonly firstName: string,
        public readonly lastName: string,
        public readonly pets: Pet[]
    ) {
    }
}
