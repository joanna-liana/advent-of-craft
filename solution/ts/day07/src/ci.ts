export interface Config {
    sendEmailSummary(): boolean;
}

export interface Emailer {
    send(message: string): void;
}

export interface Logger {
    info(message: string): void;

    error(message: string): void;
}

export enum TestStatus {
    NoTests,
    PassingTests,
    FailingTests,
}