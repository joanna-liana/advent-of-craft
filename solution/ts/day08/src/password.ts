import {Either, left, right} from 'fp-ts/Either';
import {pipe} from 'fp-ts/function';
import * as O from "fp-ts/Option";
import {none, Option, some} from "fp-ts/Option";

export class Password {
    private constructor(private readonly value: string) {
    }

    static from(input: string): Either<ParsingError, Password> {
        const rules: Array<(input: string) => Option<ParsingError>> = [
            input => Password.matches(input, /.{8,}/, 'Too short'),
            input => Password.matches(input, /.*[A-Z].*/, 'No capital letter'),
            input => Password.matches(input, /.*[a-z].*/, 'No lower letter'),
            input => Password.matches(input, /.*[0-9].*/, 'No number'),
            input => Password.matches(input, /.*[.*#@$%&].*/, 'No special character'),
            input => Password.matches(input, /^[a-zA-Z0-9.*#@$%&]+$/, 'Invalid character')
        ];

        return pipe(
            rules,
            rule => rule.map(rule => rule(input)),
            results => results.find(parsingError => O.isSome(parsingError)) ?? none,
            option => pipe(
                option,
                O.match(
                    () => right(new Password(input)),
                    onError => left(onError)
                )
            )
        );
    }

    private static matches = (input: string, regex: RegExp, reason: string): Option<ParsingError> =>
        regex.test(input)
            ? none
            : some(ParsingError.from(reason));
    toString = (): string => this.value;
}

class ParsingError {
    constructor(public reason: string) {
    }

    static from(reason: string): ParsingError {
        return new ParsingError(reason);
    }
}