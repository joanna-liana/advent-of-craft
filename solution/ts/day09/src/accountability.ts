export class Client {
    constructor(private readonly orderLines: Record<string, number>) {
    }

    statement = (): string => `${Object.entries(this.orderLines)
        .map(
            ([key, value]) => this.formatLine(key, value)
        ).join('\n')}\n${this.formatTotal()}`;

    totalAmount = (): number => Object
        .values(this.orderLines)
        .reduce((sum, value) => sum + value, 0);

    private formatLine = (name: string, value: number): string => `${name} for ${this.formatAmount(value)}`;
    private formatTotal = (): string => `Total : ${this.formatAmount(this.totalAmount())}`;
    private formatAmount = (amount: number): string => `${amount.toFixed(2)}â‚¬`;
}