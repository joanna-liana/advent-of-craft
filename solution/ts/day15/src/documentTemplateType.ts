import {RecordType} from "./recordType";

export class DocumentTemplateType {
    constructor(readonly documentType: string, readonly recordType: RecordType) {
        this.documentType = documentType;
        this.recordType = recordType;
    }
}

export class DocumentTemplates {
    public static SPEC: DocumentTemplateType = new DocumentTemplateType('SPEC', RecordType.ALL);
    public static values: Map<string, DocumentTemplateType> = new Map([
        ['DEERPP', new DocumentTemplateType('DEER', RecordType.INDIVIDUAL_PROSPECT)],
        ['DEERPM', new DocumentTemplateType('DEER', RecordType.LEGAL_PROSPECT)],
        ['AUTP', new DocumentTemplateType('AUTP', RecordType.INDIVIDUAL_PROSPECT)],
        ['AUTM', new DocumentTemplateType('AUTM', RecordType.LEGAL_PROSPECT)],
        ['SPEC', DocumentTemplates.SPEC],
        ['GLPP', new DocumentTemplateType('GLPP', RecordType.INDIVIDUAL_PROSPECT)],
        ['GLPM', new DocumentTemplateType('GLPM', RecordType.LEGAL_PROSPECT)]
    ]);

    public static fromDocumentTypeAndRecordType(documentType: string, recordType: string): DocumentTemplateType {
        let recordName = RecordType[recordType as keyof RecordType];
        let result = this.mapping()[this.formatKey(documentType, recordName)];

        if (!result) throw new Error("Invalid Document template type or record type");

        return result;
    }

    private static mapping() {
        let mapping: Map<string, DocumentTemplateType> = new Map();

        DocumentTemplates.values.forEach((dtt) => mapping[this.formatKey(dtt.documentType, dtt.recordType)] = dtt);
        Object.values(RecordType).filter((item) => isNaN(Number(item)))
            .forEach(recordType =>
                mapping[this.formatKey('SPEC', recordType)] = DocumentTemplates.SPEC
            );

        return mapping;
    }

    private static formatKey = (documentType: string, recordType: string): string =>
        `${documentType.toUpperCase()}-${recordType.toUpperCase()}`;
}