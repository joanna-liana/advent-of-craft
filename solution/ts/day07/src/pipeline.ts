import {Config, Emailer, Logger} from "./ci";
import {Project} from "./project";
import * as O from 'fp-ts/Option';
import {Option, some} from 'fp-ts/Option';
import {pipe} from "fp-ts/function";

const success = 'success';

export class Pipeline {
    constructor(private config: Config, private emailSender: Emailer, private log: Logger) {
    }

    run(project: Project): void {
        pipe(
            this.createProjectContext(project),
            O.chain(context => this.runTests(context)),
            O.chain(context => this.deploy(context))
        );
    }

    private createProjectContext(project: Project): Option<PipelineContext> {
        if (!project.hasTests()) {
            this.log.info('No tests');
        }
        return some(new PipelineContext(project));
    }

    private runTests = (context: PipelineContext): Option<PipelineContext> =>
        context.hasTests
            ? this.runTestsSafely(context)
            : some(context);

    private runTestsSafely = (context: PipelineContext): Option<PipelineContext> => some(
        this.logTestResult(
            context.withTestsResults(context.project.runTests() === success)
        )
    );

    private logTestResult(context: PipelineContext): PipelineContext {
        if (!context.testsRanSuccessfully) {
            this.log.error('Tests failed');
            this.sendEmail('Tests failed');
        } else {
            this.log.info('Tests passed');
        }
        return context;
    }

    private deploy = (context: PipelineContext): Option<PipelineContext> =>
        context.mustRunDeployment()
            ? this.deploySafely(context)
            : some(context);

    private deploySafely = (context: PipelineContext): Option<PipelineContext> => some(
        this.logDeploymentResult(
            context.withDeploymentStatus(context.project.deploy() === success)
        )
    );

    private logDeploymentResult(context: PipelineContext): PipelineContext {
        if (!context.deployedSuccessfully) {
            this.log.error('Deployment failed');
            this.sendEmail('Deployment failed');
        } else {
            this.log.info('Deployment successful');
            this.sendEmail('Deployment completed successfully');
        }
        return context;
    }

    private sendEmail(text: string): void {
        if (!this.config.sendEmailSummary()) {
            this.log.info('Email disabled');
            return;
        }
        this.log.info('Sending email');
        this.emailSender.send(text);
    }
}

class PipelineContext {
    public hasTests = this.project.hasTests();
    public testsRanSuccessfully = false;
    public deployedSuccessfully = false;

    constructor(public readonly project: Project) {
    }

    mustRunDeployment(): boolean {
        return this.testsRanSuccessfully || !this.hasTests;
    }

    withTestsResults(testsResult: boolean): PipelineContext {
        this.testsRanSuccessfully = testsResult
        return this
    }

    withDeploymentStatus(deploymentStatus: boolean): PipelineContext {
        this.deployedSuccessfully = deploymentStatus
        return this
    }
}