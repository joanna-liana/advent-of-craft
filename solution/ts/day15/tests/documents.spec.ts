import {beforeAll, describe, test} from "@jest/globals";
import {configure} from "approvals";
import {JestReporter} from "approvals/lib/Providers/Jest/JestReporter";
import {verifyAllCombinations2} from "approvals/lib/Providers/Jest/CombinationApprovals";
import {RecordType} from "../src/recordType";
import {DocumentTemplates} from "../src/documentTemplateType";

describe("documents", () => {
    beforeAll(() => {
        configure({
            reporters: [new JestReporter()],
        });
    });

    test("golden master for refactoring", () => {
        verifyAllCombinations2(
            (dtt, rt) => DocumentTemplates.fromDocumentTypeAndRecordType(dtt, rt).documentType,
            documentTemplates,
            recordTypes
        )
    });
});

const recordTypes: string[] = Object.keys(RecordType).filter((item) => {
    return isNaN(Number(item));
});

const documentTemplates: string[] = Array.from(DocumentTemplates.values).map(pair => pair[0]);