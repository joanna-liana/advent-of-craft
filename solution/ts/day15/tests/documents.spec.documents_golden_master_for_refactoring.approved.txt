[DEERPP,INDIVIDUAL_PROSPECT] => Error: Invalid Document template type or record type
[DEERPP,LEGAL_PROSPECT] => Error: Invalid Document template type or record type
[DEERPP,ALL] => Error: Invalid Document template type or record type
[DEERPM,INDIVIDUAL_PROSPECT] => Error: Invalid Document template type or record type
[DEERPM,LEGAL_PROSPECT] => Error: Invalid Document template type or record type
[DEERPM,ALL] => Error: Invalid Document template type or record type
[AUTP,INDIVIDUAL_PROSPECT] => AUTP
[AUTP,LEGAL_PROSPECT] => Error: Invalid Document template type or record type
[AUTP,ALL] => Error: Invalid Document template type or record type
[AUTM,INDIVIDUAL_PROSPECT] => Error: Invalid Document template type or record type
[AUTM,LEGAL_PROSPECT] => AUTM
[AUTM,ALL] => Error: Invalid Document template type or record type
[SPEC,INDIVIDUAL_PROSPECT] => SPEC
[SPEC,LEGAL_PROSPECT] => SPEC
[SPEC,ALL] => SPEC
[GLPP,INDIVIDUAL_PROSPECT] => GLPP
[GLPP,LEGAL_PROSPECT] => Error: Invalid Document template type or record type
[GLPP,ALL] => Error: Invalid Document template type or record type
[GLPM,INDIVIDUAL_PROSPECT] => Error: Invalid Document template type or record type
[GLPM,LEGAL_PROSPECT] => GLPM
[GLPM,ALL] => Error: Invalid Document template type or record type
